packages/wdio-browserstack-service/src/insights-handler.ts

```diff
@@ -0,0 +1,471 @@
+            if (stepDetails) {
+                stepDetails.finished_at = (new Date()).toISOString()
+                stepDetails.result = result.passed ? 'PASSED' : 'FAILED'
+                stepDetails.duration = result.duration
+                stepDetails.failure = result.error ? removeAnsiColors(result.error) : result.error
+            }
+        }
+
+        this._tests[uniqueId] = testMetaData
+    }
+
+    async uploadPending () {
+        if (this.requestQueueHandler.pendingUploads <= 0) {
+            return
+        }
+
+        await Promise.race([
+            new Promise(resolve => setTimeout(resolve, 5000)),
+            new Promise(resolve => {
+                if (this.requestQueueHandler.pendingUploads == 0) {
+                    resolve
+                }
+                while (this.requestQueueHandler.pendingUploads > 0) {
+                    // waiting for any pending uploads
+                }
+            }),
+        ])
+    }
+
+    async teardown () {
+        await this.requestQueueHandler.shutdown()
+    }
+
+    /**
+     * misc methods
+     */
+
+    async browserCommand (commandType: string, args: BeforeCommandArgs & AfterCommandArgs, test?: Frameworks.Test | ITestCaseHookParameter) {
+        if (commandType == 'client:beforeCommand') {
+            this._commands[`${args.sessionId}_${args.method}_${args.endpoint}`] = args
+            return
+        }
+
+        if (!test) {
+            return
+        }
+        const identifier = this.getIdentifier(test)
+
+        // log screenshot
+        if (process.env.BS_TESTOPS_ALLOW_SCREENSHOTS == 'true' && isScreenshotCommand(args) && args.result.value) {
+            await uploadEventData([{
+                event_type: 'LogCreated',
+                logs: [{
+                    test_run_uuid: this._tests[identifier].uuid,
+                    timestamp: new Date().toISOString(),
+                    message: args.result.value,
+                    kind: 'TEST_SCREENSHOT'
+                }]
+            }], DATA_SCREENSHOT_ENDPOINT)
+        }
+
+        const dataKey = `${args.sessionId}_${args.method}_${args.endpoint}`
+        const requestData = this._commands[dataKey]
+
+        // log http request
+        const log = {
+            test_run_uuid: this._tests[identifier].uuid,
+            timestamp: new Date().toISOString(),
+            kind: 'HTTP',
+            http_response: {
+                path: requestData.endpoint,
+                method: requestData.method,
+                body: requestData.body,
+                response: args.result
+            }
+        }
+
+        const uploadData = {
+            event_type: 'LogCreated',
+            logs: [log]
+        }
+
+        const req = this.requestQueueHandler.add(uploadData)
+
+        if (req.proceed) {
+            await uploadEventData(req.data, req.url)
+        }
+    }
+
+    /*
+     * private methods
+     */
+
+    private attachHookData (context: any, hookId: string): void {
```

Comment:
Can we type `context` here?

Reply:
Not sure. Everywhere context being used is any